core/io/sfndam.h

```diff
@@ -0,0 +1,133 @@
+#pragma once
+#include <complex>
+#include <cstdint>
+#include <istream>
+#include <numeric>
+#include <ostream>
+#include <vector>
+
+/**
+ * Implementation of the Simple NDArray with Metadata (SFNDAM) format for exhanging
+ * multi-dimensional data with some metadata (a property bag).
+ *
+ * Overall the format consists of an NDArray of type T and a JSON document with metadata.
+ * The binary layout is as follows:
+ *
+ *    1. Magic bytes 'SFNDAM' followed by a uint32 with the version number.
+ *    2. uint32 indicating the data type
+ *    3. uint32 indicating the number of dimensions in the array, R
+ *    4. Vector of R uint64 values specifying the length of each dimension.
+ *       Indexes and sizes are "row first", e.g. dimensions = [z, y, x]
+ *       Total number of elements (N) are prod(dimensions).
+ *    5. uint64 indicating the length (M) of the metadata (string characters).
+ *    6. vector of N*size(data_type) bytes for the array values. Row major storage.
+ *    7. M characters of metadata.
+ */
+
+namespace sfndam {
+
+const std::uint32_t version = 1;
+const auto version_bytes = (char*)(&version);
+const char magic_bytes[] = {'S', 'F', 'N', 'D', 'A', 'M', version_bytes[0], version_bytes[1], version_bytes[2], version_bytes[3]};
+
+enum SFNDAM_DataTypes {
+  USHORT = 1,  // uint16_t
+  SHORT = 2,   // int16_t
+  UINT = 3,    // uint32_t
+  INT = 4,     // int32_t
+  FLOAT = 5,   // float
+  DOUBLE = 6,  // double
+  CXFLOAT = 7, // complex float
+  CXDOUBLE = 8 // complex double
+};
+
+template <typename T>
+uint32_t typeid_for_type();
+
+template <>
+inline uint32_t typeid_for_type<uint16_t>() { return SFNDAM_DataTypes::USHORT; }
+template <>
+inline uint32_t typeid_for_type<int16_t>() { return SFNDAM_DataTypes::SHORT; }
+template <>
+inline uint32_t typeid_for_type<uint32_t>() { return SFNDAM_DataTypes::UINT; }
+template <>
+inline uint32_t typeid_for_type<int32_t>() { return SFNDAM_DataTypes::INT; }
+template <>
+inline uint32_t typeid_for_type<float>() { return SFNDAM_DataTypes::FLOAT; }
+template <>
+inline uint32_t typeid_for_type<double>() { return SFNDAM_DataTypes::DOUBLE; }
+template <>
+inline uint32_t typeid_for_type<std::complex<float>>() { return SFNDAM_DataTypes::CXFLOAT; }
+template <>
+inline uint32_t typeid_for_type<std::complex<double>>() { return SFNDAM_DataTypes::CXDOUBLE; }
+
+template <typename T>
+struct sfndam {
+  std::vector<uint64_t> array_dimensions;
+  std::vector<T> data;
+  std::string meta;
+};
+
+template <typename T>
+void serialize(const sfndam<T>& a, std::ostream& out) {
```

Comment:
You need to check gcount(), eof() etc. to make sure the reads were successful and the stream did does not end prematurely.